### PR TITLE
Disabling non-linear integer arithmetics

### PR DESCRIPTION
**Pull request by Dspil**
Adds a flag in silicon to disable non-linear integer arithmetics when Z3 via STDIO is used.

### CODE REVIEW

File: src/main/scala/Config.scala [hunk 1]
```diff
@@ -230,6 +230,12 @@ class Config(args: Seq[String]) extends SilFrontendConfig(args, "Silicon") {
     noshort = true
   )
 
+  val disableNLRI: ScallopOption[Boolean] = opt[Boolean]("disableNLRI",
+    descr = "Disable non-linear integer arithetics (prover must be Z3)",
```
Comment (jcp19) [inline src/main/scala/Config.scala:234]:
```suggestion
    descr = "Disable non-linear integer arithmetic when using Z3",
```

File: src/main/scala/Config.scala [hunk 2]
```diff
@@ -813,6 +819,11 @@ class Config(args: Seq[String]) extends SilFrontendConfig(args, "Silicon") {
       sys.error(s"Unexpected combination: $other")
   }
 
+  validateOpt(disableNLRI, prover) {
+    case (Some(true), n) if n != Some(Z3ProverStdIO.name) => Left(s"Option ${disableNLRI.name} is only supported with prover ${Z3ProverStdIO.name}")
```
Comment (jcp19) [inline src/main/scala/Config.scala:823]:
```suggestion
    case (Some(true), n) if n != Some(Z3ProverStdIO.name) =>
        Left(s"Option ${disableNLRI.name} is only supported with prover ${Z3ProverStdIO.name}")
```

File: src/main/scala/decider/Z3ProverStdIO.scala [hunk 3]
```diff
@@ -65,5 +65,12 @@ class Z3ProverStdIO(uniqueId: String,
     Paths.get(Verifier.config.z3Exe)
   }
 
-  override def emitSettings(contents: Iterable[String]): Unit = emit(contents)
+  override def emitSettings(contents: Iterable[String]): Unit = {
+    emit(contents)
+    if(Verifier.config.disableNLRI.getOrElse(false)) {
+      comment("We disable non-linear integer arithmetics")
```
Comment (jcp19) [inline src/main/scala/decider/Z3ProverStdIO.scala:71]:
```suggestion
      comment("We disable non-linear integer arithmetic")
```

**Review by jcp19 — COMMENTED** [review]
I have a few minor comments that we should address before merging. Could we also add a test that typically passes, but which should fail with the new flag to make sure that the implementation has the expected behaviour?

File: src/test/resources/misc/disableNL.vpr [hunk 4]
```diff
@@ -0,0 +1,7 @@
+method f(x: Int, y: Int) returns (res: Int)
+  requires 0 <= x <= 10
+  requires 0 <= y <= 10
+  ensures 0 <= res <= 100 // this fails with --disableNL
+{
+  res := x * y
+}
```
[the remark this answers is not in the document]
Comment (Dspil) [reply src/test/resources/misc/disableNL.vpr:7]: Similar tests about more complete exhale just had a comment there too

**Comment by marcoeilers** [comment]
We already have a flag to pass options to Z3, right? Is this just tomake it more convenient?

**Review by jcp19 — APPROVED** [review]
lgtm

**Pull request merged**